paddle/platform/profiler.h

```diff
@@ -0,0 +1,197 @@
+  int device_ = -1;
+#endif
+};
+
+struct EventList {
+  constexpr static std::size_t kMB = 1024 * 1024;
+  constexpr static std::size_t kEventBlockSize = 16 * kMB;
+  constexpr static std::size_t kEventSize = sizeof(Event);
+  constexpr static std::size_t kEventAlign = alignof(Event);
+  constexpr static std::size_t kNumBlock =
+      kEventBlockSize /
+      ((kEventSize + kEventAlign - 1) / kEventAlign * kEventAlign);
+
+  template <typename... Args>
+  void Record(Args&&... args) {
+    if (event_blocks.empty() || event_blocks.front().size() == kNumBlock) {
+      event_blocks.emplace_front();
+      event_blocks.front().reserve(kNumBlock);
+    }
+    event_blocks.front().emplace_back(std::forward<Args>(args)...);
+  }
+
+  std::vector<Event> Reduce() {
+    std::vector<Event> result;
+    for (auto& block : event_blocks) {
+      result.insert(result.begin(), std::make_move_iterator(block.begin()),
+                    std::make_move_iterator(block.end()));
+    }
+    event_blocks.clear();
+    return result;
+  }
+
+  std::forward_list<std::vector<Event>> event_blocks;
+};
+
+enum ProfilerState {
+  kDisabled,
+  kCPU,
+  kCUDA,
+};
+
+// The profiler state, the initial value is ProfilerState::kDisabled
+extern ProfilerState kState;
+// The global mutex
+extern std::mutex kAllEventListsMutex;
+// The total event lists of all threads
+extern std::list<std::shared_ptr<EventList>> kAllEventLists;
+// The thread local event list only can be accessed by the specific thread
+extern thread_local std::shared_ptr<EventList> kEventList;
+// The thread index of each thread
+extern thread_local int32_t kThreadId;
+// The kNextThreadId is a global counter for threads, by the kThreadId and
+// kNextThreadId, we can know how many threads have created EventList.
+extern uint32_t kNextThreadId;
+
+inline EventList& GetEventList() {
+  if (!kEventList) {
+    std::lock_guard<std::mutex> guard(kAllEventListsMutex);
+    kEventList = std::make_shared<EventList>();
+    kThreadId = kNextThreadId++;
+    kAllEventLists.emplace_front(kEventList);
+  }
+  return *kEventList;
+}
+
+inline void Mark(const std::string name,
+                 const platform::DeviceContext* dev_ctx = nullptr) {
+  GetEventList().Record(EventKind::kMark, std::move(name), kThreadId, dev_ctx);
+}
+
+struct RecordEvent {
+  explicit RecordEvent(const std::string name,
+                       platform::DeviceContext* dev_ctx = nullptr) {
+    if (kState == ProfilerState::kDisabled) return;
+    dev_ctx_ = dev_ctx;
+    GetEventList().Record(EventKind::kPushRange, std::move(name), kThreadId,
+                          dev_ctx_);
+  }
+
+  ~RecordEvent() {
+    if (kState == ProfilerState::kDisabled) return;
+    GetEventList().Record(EventKind::kPopRange, std::string(), kThreadId,
+                          dev_ctx_);
+  }
+  platform::DeviceContext* dev_ctx_;
+};
+
+void EnableProfiler(ProfilerState state);
+std::vector<std::vector<Event>> DisableProfiler();
```

Comment:
Need more comments for return format of `DisableProfiler`

Reply:
Done